### PR TITLE
Actions2functions

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?

----

## Description of change

Add functions support (charm.v6 update and hook symlinks);

depends on:

- https://github.com/juju/charm-tools/pull/555
- https://github.com/juju/charm/pull/299

Driveby fix bootstrap to k8s: pod not found  issue:

```
ERROR juju.cmd.juju.commands bootstrap.go:778 failed to bootstrap model: creating controller stack for controller: creating statefulset for controller: fetching pods' status for controller: pod not found not found
```

### CaaS

```console
# deploy charm has `functions.yaml` and `functions/`;

# check operator hooks symlinks;
$ root@mariadb-k8s-0:/var/lib/juju# ll tools/unit-mariadb-k8s-0/action-*
lrwxrwxrwx 1 root root 25 Nov 15 07:41 tools/unit-mariadb-k8s-0/action-fail -> /var/lib/juju/tools/jujud*
lrwxrwxrwx 1 root root 25 Nov 15 07:41 tools/unit-mariadb-k8s-0/action-get -> /var/lib/juju/tools/jujud*
lrwxrwxrwx 1 root root 25 Nov 15 07:41 tools/unit-mariadb-k8s-0/action-log -> /var/lib/juju/tools/jujud*
lrwxrwxrwx 1 root root 25 Nov 15 07:41 tools/unit-mariadb-k8s-0/action-set -> /var/lib/juju/tools/jujud*
$ root@mariadb-k8s-0:/var/lib/juju# ll tools/unit-mariadb-k8s-0/function-*
lrwxrwxrwx 1 root root 25 Nov 15 07:41 tools/unit-mariadb-k8s-0/function-fail -> /var/lib/juju/tools/jujud*
lrwxrwxrwx 1 root root 25 Nov 15 07:41 tools/unit-mariadb-k8s-0/function-get -> /var/lib/juju/tools/jujud*
lrwxrwxrwx 1 root root 25 Nov 15 07:41 tools/unit-mariadb-k8s-0/function-log -> /var/lib/juju/tools/jujud*
lrwxrwxrwx 1 root root 25 Nov 15 07:41 tools/unit-mariadb-k8s-0/function-set -> /var/lib/juju/tools/jujud*

# list functions;
$ juju functions mariadb-k8s
Function  Description
hello     Hello World.

# show functions;
$ juju show-function mariadb-k8s hello
Hello World.

Arguments
who:
  type: string
  description: The person to say hello to.

# call function
$ juju call mariadb-k8s/0 hello who=kelvin
Running Task 1
Waiting for task 1...
outcome: success
result-map:
  message: |-
    Hello
    kelvin!
  time-completed: Fri Nov 15 08:35:06 UTC 2019

$ juju tasks --apps mariadb-k8s
Time                 Id  Task   Status     Unit
2019-11-15 19:35:07   1  hello  completed  mariadb-k8s/0

$ juju show-task 1
outcome: success
result-map:
  message: |-
    Hello
    kelvin!
  time-completed: Fri Nov 15 08:35:06 UTC 2019
```

### IaaS

```console
# bootstrap to aws;
$ juju bootstrap aws/ap-southeast-2 aws --debug --build-agent

# download postgresql charm and rename `actions` folder to `functions`, `actions.yaml` to `functions.yaml`
# build charm then deploy;
$ charm-build ./ --no-local-layers  && juju deploy /tmp/charm-builds/postgresql/ --debug

$ juju functions postgresql
Function            Description
replication-pause   Pause replication replay on a hot standby unit.
replication-resume  Resume replication replay on a hot standby unit.
switchover          Promote a specific unit to master. Must be run on the leader unit.
test-function       fake echo function.
wal-e-backup        Run a wal-e backup to cloud storage now. Requires WAL shipping to be enabled with wal-e. Action terminates when the backup is complete.
wal-e-list-backups  List backups available for PITR and their metadata.
wal-e-restore       PITR database recovery from configured wal-e store. THIS WILL DESTROY YOUR EXISTING DATA. Most of these options correspond to PostgreSQL recovery target settings, documented at http://www.postgresql.org/docs/current/static/recovery-target-settings.html.

$ juju show-function postgresql test-function
fake echo function.

Arguments
arg:
  type: string
  description: null

$ juju call postgresql/0 test-function arg=hey
Running Task 3
Waiting for task 3...
result: "printed params -> \n{'arg': 'hey'}"

$ juju show-task 3
result: "printed params -> \n{'arg': 'hey'}"

$ juju tasks
Time                 Id  Task           Status     Unit
2019-11-18 15:18:44   1  test-function  completed  postgresql/0
2019-11-18 15:31:15   2  test-function  completed  postgresql/0
2019-11-18 15:31:25   3  test-function  completed  postgresql/0

$ juju run --unit postgresql/0 --timeout=30s env | grep -i "function\|action"
JUJU_ACTION_UUID=7
JUJU_ACTION_NAME=juju-run
JUJU_FUNCTION_TAG=action-7
JUJU_ACTION_TAG=action-7
JUJU_FUNCTION_NAME=juju-run
JUJU_FUNCTION_ID=7
```

## Documentation changes

No

## Bug reference

None

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -147,7 +148,9 @@
 
 [[constraint]]
   name = "gopkg.in/juju/charm.v6"
-  revision = "780719c3e4a68329408cfecf46ed1ee84b530287"
+  # revision = "780719c3e4a68329408cfecf46ed1ee84b530287"
```
Comment (ycliuhw) [inline Gopkg.toml]: Update charm.v6 sha once it's landed!

File: Gopkg.toml [hunk 2]
```diff
@@ -24,6 +24,7 @@
 #   go-tests = true
 #   unused-packages = true
 
+
```
Comment (wallyworld) [inline Gopkg.toml]: delete me

File: worker/uniter/runner/runner.go [hunk 3]
```diff
@@ -277,7 +277,7 @@ func (runner *runner) updateActionResults(results *utilexec.ExecResponse) error
 }
 
 // RunAction exists to satisfy the Runner interface.
-func (runner *runner) RunAction(actionName string) error {
+func (runner *runner) RunAction(actionName string) (err error) {
```
Comment (wallyworld) [inline worker/uniter/runner/runner.go]: why this change?

File: worker/uniter/runner/runner.go [hunk 4]
```diff
@@ -286,10 +286,23 @@ func (runner *runner) RunAction(actionName string) error {
...
-	// run actions on remote workload pod for caas.
-	return runner.runCharmHookWithLocation(actionName, "actions", rMode)
+	dirName := "actions"
+	if runner.isFunction() {
+		dirName = "functions"
+	}
+	return runner.runCharmHookWithLocation(actionName, dirName, rMode)
+}
+
+func (runner *runner) isFunction() bool {
```
Comment (wallyworld) [inline worker/uniter/runner/runner.go]: We should just return the dir name to use here.

```
dirName := functionDir()
return runner.runCharmHookWithLocaton(actionName, dirName, rMode)
```

**Comment by wallyworld** [comment]
We also need another change added.

There is this code:

```
	if context.actionData != nil {
		vars = append(vars,
			"JUJU_ACTION_NAME="+context.actionData.Name,
			"JUJU_ACTION_UUID="+context.actionData.Tag.Id(),
			"JUJU_ACTION_TAG="+context.actionData.Tag.String(),
		)
	}

```

We need to set:
JUJU_FUNCTION_NAME
JUJU_FUNCTION_ID
JUJU_FUNCTION_TAG

note ID not UUID for function

File: worker/uniter/runner/runner.go (hunk 3 above)
Comment (ycliuhw) [reply worker/uniter/runner/runner.go]: removed

**Comment by ycliuhw** [comment]
[fix the gap from charm buid tool to Juju for functions](https://trello.com/c/1ptSj8mp/978-fix-the-gap-from-charm-buid-tool-to-juju-for-functions)

**Comment by ycliuhw** [comment]
$$merge$$
